geth/peers/syncmode.go

```diff
@@ -0,0 +1,108 @@
+package peers
+
+import (
+	"sync"
+	"time"
+)
+
+type syncStrategy struct {
+	fastMode      time.Duration
+	slowMode      time.Duration
+	fastModeLimit time.Duration
+
+	mode        chan time.Duration // for internal usage
+	currentMode time.Duration      // can only be read or write in Start() goroutine
+	period      chan time.Duration // deduped and returned by Start()
+
+	mu   sync.RWMutex
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+func newSyncStrategy(fastMode, slowMode, fastModeLimit time.Duration) *syncStrategy {
+	return &syncStrategy{
+		fastMode:      fastMode,
+		slowMode:      slowMode,
+		fastModeLimit: fastModeLimit,
+	}
+}
+
+// limitFastMode switches to slow mode sync after certain amount of time.
+// If the timer already exists, it's canceled.
+func (s *syncStrategy) limitFastMode(timeout time.Duration, cancel <-chan struct{}) {
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		select {
+		case <-time.After(timeout):
+			s.mode <- s.slowMode
+		case <-cancel:
+		}
+	}()
+}
+
+func (s *syncStrategy) Start() <-chan time.Duration {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.quit = make(chan struct{})
+	// use buffered channel so it does not block on initialization
+	s.period = make(chan time.Duration, 2)
+	s.mode = make(chan time.Duration, 2)
+
+	s.wg.Add(1)
+	go func() {
```

Comment:
Any reason to not create a method for this goroutine?

```
func (s *syncStrategy) loop() { ... }

func (s *syncStrategy) Start() <-chan time.Duration {
    ...
    go s.loop()
    ...
}
```

Les indentation and more clear.